// supabase-cli@0.0.21/dist/commands/migration/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = exports.action = void 0;
const fs_1 = require("fs");
const commander_1 = require("commander");
const database_1 = require("../../libs/database");
const action = async (name, { tableName }) => {
    await fs_1.promises.mkdir("supabase/migrations").catch(() => undefined);
    const fileName = Intl.DateTimeFormat("us", {
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
        hour: "2-digit",
        minute: "2-digit",
        second: "2-digit",
    })
        .format(new Date())
        .replace(/[/: ]/g, "") +
        (name ? `_${name}` : tableName ? `_${tableName}` : "");
    const filePath = `supabase/migrations/${fileName}.sql`;
    if (tableName) {
        (0, database_1.dumpTable)({ fileName: filePath, tableName });
    }
    else
        fs_1.promises.writeFile(filePath, "");
};
exports.action = action;
exports.create = commander_1.program
    .createCommand("create")
    .description("Create migration")
    .addArgument(new commander_1.Argument("[name]", "Migration name"))
    .addOption(new commander_1.Option("-t, --tableName <tableName>", "Output table name"))
    .action(exports.action);
