// supabase-cli@0.0.21/dist/enums/infrastructure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_PROJECT_API_SERVICE_ID = exports.DEFAULT_FREE_PROJECTS_LIMIT = exports.PASSWORD_STRENGTH_PERCENTAGE = exports.PASSWORD_STRENGTH_COLOR = exports.PASSWORD_STRENGTH = exports.DEFAULT_MINIMUM_PASSWORD_STRENGTH = exports.STRIPE_PRODUCT_IDS = exports.PROJECT_STATUS = exports.SERVICE_STATUS = exports.AWS_SUPPORTED_AZ = exports.PROVIDERS = exports.PRICING_TIER_DEFAULT_KEY = exports.PRICING_TIER_FREE_KEY = exports.PRICING_TIER_PRODUCT_IDS = exports.PRICING_TIER_LABELS = exports.REGIONS_DEFAULT = exports.REGIONS = void 0;
// Alias regions remain as the starting point for project creation
// they are immediately translated to their respective cloud regions
// and are afterward never referred to
exports.REGIONS = {
    WEST_US: "West US (North California)",
    EAST_US: "East US (North Virginia)",
    CENTRAL_CANADA: "Canada (Central)",
    WEST_EU: "West EU (Ireland)",
    WEST_EU_2: "West EU (London)",
    // 'North EU': 'North EU',
    CENTRAL_EU: "Central EU (Frankfurt)",
    SOUTH_ASIA: "South Asia (Mumbai)",
    SOUTHEAST_ASIA: "Southeast Asia (Singapore)",
    NORTHEAST_ASIA: "Northeast Asia (Tokyo)",
    NORTHEAST_ASIA_2: "Northeast Asia (Seoul)",
    OCEANIA: "Oceania (Sydney)",
    SOUTH_AMERICA: "South America (São Paulo)",
    // SOUTH_AFRICA: 'South Africa (Cape Town)',
};
exports.REGIONS_DEFAULT = exports.REGIONS.EAST_US;
exports.PRICING_TIER_LABELS = {
    FREE: "Free",
    PRO: "Pro",
};
exports.PRICING_TIER_PRODUCT_IDS = {
    FREE: "tier_free",
    PRO: "tier_pro",
    PAYG: "tier_payg",
    ENTERPRISE: "tier_enterprise",
};
exports.PRICING_TIER_FREE_KEY = "FREE";
exports.PRICING_TIER_DEFAULT_KEY = "FREE";
exports.PROVIDERS = {
    AWS: {
        id: "AWS",
        name: "aws",
        DEFAULT_SSH_KEY: "supabase-app-instance",
        regions: {
            EAST_US: "us-east-1",
            // 'North EU': 'eu-north-1', // Stockholm
            WEST_US: "us-west-1",
            SOUTHEAST_ASIA: "ap-southeast-1",
            NORTHEAST_ASIA: "ap-northeast-1",
            NORTHEAST_ASIA_2: "ap-northeast-2",
            OCEANIA: "ap-southeast-2",
            WEST_EU: "eu-west-1",
            WEST_EU_2: "eu-west-2",
            CENTRAL_EU: "eu-central-1",
            CENTRAL_CANADA: "ca-central-1",
            SOUTH_ASIA: "ap-south-1",
            SOUTH_AMERICA: "sa-east-1", // Sao Paulo
            // SOUTH_AFRICA: 'af-south-1', // Cape Town
        },
        sizes: {
            nano: {
                x86_64: "t3.nano",
                arm64: "t4g.nano",
                description: "512mb RAM",
            },
            micro: {
                x86_64: "t3.micro",
                arm64: "t4g.micro",
                description: "1gb RAM",
            },
            medium: {
                x86_64: "m5a.large",
                arm64: "m6g.medium",
                description: "4gb RAM",
            },
        },
    },
};
// @todo ini update for prod
exports.AWS_SUPPORTED_AZ = {
    EAST_US: [
        "us-east-1a",
        "us-east-1b",
        "us-east-1c",
        "us-east-1d",
        "us-east-1e",
        "us-east-1f",
    ],
    SOUTHEAST_ASIA: ["ap-southeast-1a", "ap-southeast-1b", "ap-southeast-1c"],
    WEST_EU: ["eu-west-1a", "eu-west-1b", "eu-west-1c"],
};
exports.SERVICE_STATUS = {
    INACTIVE: "INACTIVE",
    ACTIVE_HEALTHY: "ACTIVE_HEALTHY",
    ACTIVE_UNHEALTHY: "ACTIVE_UNHEALTHY",
    COMING_UP: "COMING_UP",
    UNKNOWN: "UNKNOWN",
    GOING_DOWN: "GOING_DOWN",
    INIT_FAILED: "INIT_FAILED",
    REMOVED: "REMOVED",
};
exports.PROJECT_STATUS = {
    INACTIVE: "INACTIVE",
    ACTIVE_HEALTHY: "ACTIVE_HEALTHY",
    ACTIVE_UNHEALTHY: "ACTIVE_UNHEALTHY",
    COMING_UP: "COMING_UP",
    UNKNOWN: "UNKNOWN",
    GOING_DOWN: "GOING_DOWN",
    INIT_FAILED: "INIT_FAILED",
    REMOVED: "REMOVED",
    RESTORING: "RESTORING",
};
exports.STRIPE_PRODUCT_IDS = {
    FREE: process?.env?.NEXT_PUBLIC_STRIPE_FREE_TIER_ID || "prod_Ip4vqwv3EJ7Mi0",
    PRO: process?.env?.NEXT_PUBLIC_STRIPE_PRO_TIER_ID || "prod_IsRLOp58Z7V4XN",
    PAYG: process?.env?.NEXT_PUBLIC_STRIPE_PAYG_TIER_ID || "prod_JlTbw91xcM6NY4",
};
exports.DEFAULT_MINIMUM_PASSWORD_STRENGTH = 4;
exports.PASSWORD_STRENGTH = {
    0: "This password is not acceptable.",
    1: "This password is not secure enough.",
    2: "This password is not secure enough.",
    3: "Not bad, but your password must be harder to guess.",
    4: "This password is strong.",
};
exports.PASSWORD_STRENGTH_COLOR = {
    0: "bg-red-900",
    1: "bg-red-900",
    2: "bg-yellow-900",
    3: "bg-yellow-900",
    4: "bg-green-900",
};
exports.PASSWORD_STRENGTH_PERCENTAGE = {
    0: "10%",
    1: "30%",
    2: "50%",
    3: "80%",
    4: "100%",
};
exports.DEFAULT_FREE_PROJECTS_LIMIT = 2;
exports.DEFAULT_PROJECT_API_SERVICE_ID = 1;
